### ISSUE TITLE
Progressively enhanced HLS support

### ISSUE THREAD

**Issue by mihar-22** [vidstack/vidstack#376]
**TLDR;** Currently `HlsElement` + `hls.js` delays first contentful paint (FCP) by `~730ms`, of which `75%` is contributed by `hls.js`. Avoiding it in the initial render path can lead to a reduced delay of `~180ms`, amounting to a reduction of `~550ms`. This is huge as the total `FCP` should be below the recommended `1.8s`, this amounts to `30%` of it. Splitting it out of the main bundle will also mean it can be cached separately leading to additional performance gains.

## Overview

Today the [`hls.js`](https://github.com/video-dev/hls.js) library is directly bundled with the `HlsElement` (our HLS media provider) to enable using the [HLS streaming protocol](https://en.wikipedia.org/wiki/HTTP_Live_Streaming) for delivering media. This can incur significant performance overhead on initial application rendering due to the sheer size of the library. Bundlephobia [reports](https://bundlephobia.com/package/hls.js@1.0.7) that it comes in at `240kb` min and `71kb` minzip with no treeshaking support on version `0.14.7` (currently supported). These numbers only get worse post `1.0`.

Logically it makes sense to avoid loading `hls.js` where HLS is ["natively supported"](https://caniuse.com/?search=hls) but generally doesn't work out that simple. This needs to be explored a little deeper but we've noticed certain videos stop loading, playback issues and inconsistency across Safari versions. Furthermore, I know it gets way more painful once you enter any kind of captions support. So it's not as simple as dynamically loading `hls.js` based on browser support.

Thus, we'll roll out a feature for progressively enhancing/supporting HLS playback and only loading `hls.js` when needed. Browser support will only be one way to determine loading it (but not the only). We'll need to determine an appropriate set of errors and feature requirements to look at when determining when to "upgrade".

## Solution Stages

- [x] **STAGE 1:** In the first stage we'll simply load `hls.js` dynamically and not expect it to be bundled. This way we'll get it out of the initial render path and allow it to be cached separately. You can still opt to bundle it by passing it through the `hlsLibrary` property on the `HlsElement`, or you can pass in a `string` to load it in from a CDN/Server. By default it'll point to the [JSDelivr](https://jsdelivr.com) CDN.
- [ ] **STAGE 2:** In the second stage we'll need to determine the appropriate errors and feature requirements to determine when to "upgrade" to `hls.js`.

## Performance Profile

**The following is not a thorough analysis but an effort to get some baseline metrics.**

**NETWORK:** Fast 3G
**CPU:** 4x slowdown
**BUNDLE:** I'm testing purely with the `HlsElement` component (included in`vds*.js`) and code splitting external dependencies to get an accurate reading.

- `vds` bundle -> `~150ms` to download and `~30ms` to evaluate.
- `hls.js` bundle -> `~500ms` to download and `~50ms` to evaluate.

In total we're delaying the `FCP` by `~730ms`, of which `75%` (`~550ms`) is contributed by `hls.js`. At best avoiding `hls.js` on initial render path can lead us to a reduced `FCP` delay of `~180ms` for solely the `HlsElement` (not including core dependencies such as `lit` or `tslib` which would be shared at the application level). 

### Network

⚠️ TTFB (green bar) is irrelvant here. Only content download time matters (blue bar).

![image](https://user-images.githubusercontent.com/14304599/125779849-7219e8b9-8553-4f00-897b-8bd8ebba18c8.png)

### `HlsElement` + `hls.js` evaluation
![image](https://user-images.githubusercontent.com/14304599/125778937-0d32c083-c77f-435d-b8d8-f022b4be1d64.png)

### `hls.js` evaluation

![image](https://user-images.githubusercontent.com/14304599/125778995-6c1ee123-7fe6-4aad-af04-b0e85b54fa95.png)
